plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+--     instance (ExampleUni `Contains` a, ExampleUni `Contains` b) => ExampleUni `Contains` (a, b) where <...>
+--
+-- (a universe can have any subset of the mentioned sorts of types, for example it's fine to have
+-- instantiated polymorphic types and not have uninstantiated ones and vice versa)
+--
+-- Note that when used as a constraint of a function 'Contains' does not allow you to directly
+-- express things like \"@uni@ has the @Integer@, @Bool@ and @[]@ types and type formers\",
+-- because @[]@ is not fully instantiated. So you can only say \"@uni@ has @Integer@, @Bool@,
+-- @[Integer]@, @[Bool]@, @[[Integer]]@, @[[Bool]]@ etc\" and such manual enumeration is annoying,
+-- so we'd really like to be able to say that @uni@ has lists of arbitrary built-in types
+-- (including lists of lists etc). 'Contains' does not allow that, but 'Includes' does.
+-- For example, in the body of the following definition:
+--
+--     foo :: (uni `Includes` Integer, uni `Includes` Bool, uni `Includes` []) => <...>
+--     foo = <...>
+--
+-- you can make use of the fact that @uni@ has lists of arbitrary included types (integers,
+-- booleans and lists).
+--
+-- Hence most of the time opt for using the more flexible 'Includes'.
+--
+-- 'Includes' is defined in terms of 'Contains', so you only need to provide a 'Contains' instance
+-- per type from the universe and you'll get 'Includes' for free.
+type Contains :: forall k. (Type -> Type) -> k -> Constraint
+class uni `Contains` a where
+    knownUni :: uni (T a)
+
+{- Note [The definition of Includes]
+We need to be able to partially apply 'Includes' (required in the definition of '<:' for example),
+however if we define 'Includes' as a class alias like that:
+
+    class    Contains uni `Permits` a => uni `Includes` a
+    instance Contains uni `Permits` a => uni `Includes` a
+
+we get this extra annoying warning:
+
+    • The constraint ‘Includes uni ()’ matches
+        instance forall k (uni :: * -> *) (a :: k).
+                 Permits (Contains uni) a =>
+                 Includes uni a
+      This makes type inference for inner bindings fragile;
+        either use MonoLocalBinds, or simplify it using the instance
+
+at the use site, so instead we define 'Includes' as a type alias of one argument (i.e. 'Includes'
+has to be immediately applied only to a @uni@ at the use site).
+-}
+
+-- See Note [The definition of Includes].
+-- | @uni `Includes` a@ reads as \"@a@ is in the @uni@\". @a@ can be of a higher-kind,
+-- see the docs of 'Contains' on why you might want that.
+type Includes :: forall k. (Type -> Type) -> k -> Constraint
+type Includes uni = Permits (Contains uni)
+
+-- | Same as 'knownUni', but receives a @proxy@.
+knownUniOf :: uni `Contains` a => proxy a -> uni (T a)
+knownUniOf _ = knownUni
+
+-- | Wrap a value into @Some (ValueOf uni)@, given its explicit type tag.
+someValueOf :: forall a uni. uni (T a) -> a -> Some (ValueOf uni)
```

Comment:
`someTaggedValue`?

Reply:
(I find the "tag" naming very helpful, hence why I'm suggesting it all over the place)

Reply:
Even if we end up having tags in some form, I'd like to keep it `someValueOf`, 'cause it receives a type tag and a value **of** that type and literally returns a `Some (ValueOf uni)`.